pymc_marketing/mmm/base.py

```diff
@@ -390,24 +396,168 @@ def plot_contribution_curves(self) -> plt.Figure:
+            A dictionary specifying the budget bounds for each channel.
+
+        Returns
+        -------
+        pd.DataFrame
+            A DataFrame with the optimal budget allocation for each channel.
+
+        Raises
+        ------
+        ValueError
+            If any of the required parameters are not provided or have an incorrect type.
+
+        Notes
+        -----
+        - The function optimally allocates the budget among different channels based on the provided estimations and budget constraints.
+        - It utilizes a cost function that calculates the contribution for each channel based on a quadratic curve fit to the data.
+        - The budget allocation is subject to a constraint to ensure the total allocated budget matches the total budget.
+        - The function uses optimization techniques to find the optimal budget allocation that maximizes contribution.
+        - The updated DataFrame includes an 'Optimal Budget' column with the allocated budget for each channel.
+        - The function also prints the maximum total contribution with the allocated budget and the proposed budget amount.
+        """
+        
+        if total_budget is None or df_estimations is None:
+            raise ValueError("Total Budget and Estimations dataframe parameters must be provided.")
+            
+        if not isinstance(total_budget, (int, float)):
+            raise ValueError("The 'total_budget' parameter must be an integer or float.")
+            
+        if not isinstance(df_estimations, pd.DataFrame):
+            raise ValueError("The 'df_estimations' parameter must be a pandas DataFrame.")
+            
+        if not isinstance(budget_bounds, dict):
+            raise ValueError("The 'budget_bounds' parameter must be a dictionary.")
+            
+        df = df_estimations.copy()
+        channel_contributions = self.compute_channel_contribution_original_scale().mean(
+            ["chain", "draw"]
+        )
+        
+        def cost_function(budget_allocations, df):
+            total_collaboration = 0
+            for i, allocation in enumerate(budget_allocations):
+                # Retrieve the data for this channel
+                channel = df.loc[i, 'channel']
+                x = self.data[channel].to_numpy().copy()
+                y = np.array(channel_contributions.sel(channel=channel).copy())
+                
+                # Fit a quadratic curve to the data
+                # Call the new function
+                polynomial, _, _, _, _, _ = calculate_curve(x, y)
+        
+                # Calculate the collaboration for this channel based on its quadratic curve
+                total_collaboration -= polynomial(allocation)
+            return total_collaboration
+
+        def budget_constraint(budget_allocations):
+            total_budget_allocated = np.sum(budget_allocations)
+            return total_budget_allocated - total_budget  # this should be zero
```

Comment:
Maybe we can extract these functions to a module e.g. `pymc_marketing/mmm/budget_optimizer`?